### ISSUE TITLE
Add support for a function in place of an Object in plugin options

### ISSUE THREAD

**Issue by wen911119** [fastify/fastify#1489]
## 🚀 Feature Proposal
now
```javascript
fastify.register(require('./my-plugin'), opts)
```
expected
```javascript
fastify.register(require('./my-plugin'), optsOrFunctionToReturnOpts)
```

## Motivation
./plugins/fastify-config-loader.js
```javascript
module.exports = fp((fastify, options, next) => {
  fastify.decorate('config',{
    swagger: {}
  })
  next()
})
```

index.js
```javascript
fastify.register(require('./plugins/fastify-config-loader'))
// TypeError: Cannot read property 'swagger' of undefined
fastify.register(require('fastify-swagger'), fastify.config.swagger)
```
:sob: fastify.config is undefined at this moment :sob:
## Example

Please provide an example for how this feature would be used.
index.js
```javascript
fastify.register(require('./plugins/fastify-config-loader'))
fastify.register(require('fastify-swagger'), () => {
  return fastify.config.swagger
})
```
:smile:

**Comment by mcollina** [comment]
Why don't you use `.after()`? That'd work as expected.

**Comment by mcollina** [comment]
That's a cool  one! Yes, I think it would be valuable!

**Comment by alex-ppg** [comment]
I propose you change the title to "*Add support for a function in plugin opts*".

This issue was not with fastify ***directly*** but rather the [avvio](https://github.com/mcollina/avvio) plugin handler dependency which offered no support for this feature. I have created the pull request [#82](https://github.com/mcollina/avvio/pull/82) which solves this issue by introducing variable checking and deep-cloning via a native C++ module by [@scottinet](https://github.com/scottinet).

**Comment by alex-ppg** [comment]
I believe this issue is now resolved thanks to the merged [PR#1498](https://github.com/fastify/fastify/pull/1498)

**Comment by Eomm** [comment]
You have right!
Feature description [here](https://github.com/fastify/fastify/blob/master/docs/Plugins-Guide.md#how-to-handle-encapsulation-and-distribution)

Closing
